### PR TITLE
32 slurm

### LINKED ISSUE #32: Replace bash script generation with slurm
In the current implementation of #19, we have two scripts, `generate_metric_scripts.py` and `generate_train_scripts.py` which generate several bash scripts over the combination of datasets and seeds within a given experiment group. The generated bash scripts are then called one by one.

We need to switch this out to using slurm for submitting batch jobs.

### PR DESCRIPTION
**Pull request by philswatton**
This PR resolves #32 for the case of training scripts, and contributes to #19. It adds the following:

- `generate_train_scripts.py` now uses templating to generate a slurm batch submission
- `train_all` will submit all batches for a given experiment group
- model checkpointing added, to save the final result
- epochs reduced to a more sensible 20
- pytorch `seed_everything` call added
- `scripts/README` updated

What this PR *doesn't* do is #26, which is left for another PR.

### CODE REVIEW

**Comment by philswatton** [comment]
In the end, it turned out that naming the model artifacts was broken because this was only added for the wandb logger in version 1.9.0 of pytorch lightning (and indeed it seems no other way existed until this was added, see: https://github.com/Lightning-AI/lightning/issues/15990). I did try using the `filename` argument in `ModelCheckpoint` but this didn't work.

Since our project dependency is 1.8.3.post1 and updating would be a big hassle (and in particular I recall things not working on Baskerville?) I took the route of adapting the changes to the wandb logger in 1.9.0 to 1.8.3.post1 and adding it to our project source code. This does work, and I ran I couple of very quick 100-step models to check that it did.

Lmk if that's too hacky a solution!

File: scripts/train_models.py [hunk 1]
```diff
@@ -38,11 +41,13 @@ def train_model(dm: LightningDataModule, experiment_name: str, trainer_config: d
 
 def train_models(dmpair_kwargs: dict, trainer_config: dict, experiment_pair_name: str):
     dmpair = DMPair(**dmpair_kwargs)
+    seed_everything(seed=dmpair_kwargs["seed"])
     train_model(
         dm=dmpair.A,
         experiment_name=f"{experiment_pair_name}_A",
         trainer_config=trainer_config,
     )
+    seed_everything(seed=dmpair_kwargs["seed"])
```
Comment (lannelin) [inline scripts/train_models.py:50]: should the seeding happen before DMPair creation for control of val split?

File: src/modsim2/model/wandb_logger.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# The code below taken from pytorch_lightning 1.9.3 and adapted
+# to the project's version of 1.8.3.post1 to avoid updating
+# and breaking everything
```
Comment (lannelin) [inline src/modsim2/model/wandb_logger.py:3]: as discussed, let's raise an issue to make this temporary

File: configs/trainer.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 trainer_kwargs:
-  max_epochs: 100
+  max_epochs: 10
```
Comment (lannelin) [inline configs/trainer.yaml:2]: is this enough for convergence?

File: scripts/train_all.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Useage: scripts/train_all.sh drop-only
```
Comment (lannelin) [inline scripts/train_all.sh:1]: rename file to be clear we're using slurm
also typo in usage

File: scripts/templates/train-template.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
```
Comment (lannelin) [inline scripts/templates/train-template.sh:1]: rename to reflect slurm so like `slurm-train-template.sh`

File: scripts/README.md [hunk 6]
```diff
@@ -30,7 +30,13 @@ python scripts/generate_train_scripts.py --experiment_group "drop-only"
 
 to generate bash scripts in `ROOT/train_scripts/`. Note additional optional arguments include `--dataset_config_path` and `--train_config_path` (see below).
```
Comment (lannelin) [inline scripts/README.md:31]: Let's be clear about slurm usage throughout, "generate *slurm* scripts"

File: scripts/templates/train-template.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+#SBATCH --output ./slurm_logs/%j.out
+
+module purge
+module load baskerville
+module load Miniconda3/4.10.3
+export CONDA_PKGS_DIRS=/tmp
+eval "$(${EBROOTMINICONDA3}/bin/conda shell.bash hook)"
+
+# Define the path to your Conda environment (modify as appropriate)
+# N.B. this path will be created by the subsequent commands if it doesn't already exist
```
Comment (lannelin) [inline scripts/templates/train-template.sh:18]: will it?

File: scripts/templates/train-template.sh [hunk 8]
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
+#SBATCH --account vjgo8416-mod-sim-2  # Only required if you are a member of more than one Baskerville project
```
Comment (lannelin) [inline scripts/templates/train-template.sh:2]: can you remove some of the boilerplate comments here?

File: scripts/README.md (hunk 6 above)
Comment (lannelin) [reply scripts/README.md:31]: should state prereqs for using slurm script (access to slurm, conda env creation - though not necessarily specific steps required on *our* HPC to create it!)

File: src/modsim2/model/wandb_logger.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                            "save_top_k",
+                            "save_weights_only",
+                            "_every_n_train_steps",
+                        ]
+                        # ensure it does not break if `ModelCheckpoint`
+                        # args change
+                        if hasattr(checkpoint_callback, k)
+                    },
+                }
+                if _WANDB_GREATER_EQUAL_0_10_22
```
Comment (lannelin) [inline src/modsim2/model/wandb_logger.py:78]: what are we doing if not? should this raise an error?

File: src/modsim2/model/wandb_logger.py (hunk 2 above)
Comment (lannelin) [reply src/modsim2/model/wandb_logger.py:3]: link to issue 37

File: scripts/README.md (hunk 6 above)
Comment (lannelin) [reply scripts/README.md:31]: ah ok we should make it clear that this setups up slurm scripts for *our* HPC as there are some Baskerville specifics

File: scripts/templates/train-template.sh [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+module purge
+module load baskerville
+module load Miniconda3/4.10.3
+export CONDA_PKGS_DIRS=/tmp
+eval "$(${EBROOTMINICONDA3}/bin/conda shell.bash hook)"
+
+# Define the path to your Conda environment (modify as appropriate)
+# N.B. this path will be created by the subsequent commands if it doesn't already exist
+CONDA_ENV_PATH="/bask/projects/v/vjgo8416-mod-sim-2/ms2env"
```
Comment (lannelin) [inline scripts/templates/train-template.sh:19]: let's pass this in to the generation script to let this be flexible

File: scripts/templates/train-template.sh (hunk 8 above)
Comment (lannelin) [inline scripts/templates/train-template.sh:2]: pass in this val to generation script for flexibility

**Review by lannelin — CHANGES_REQUESTED** [review]
looks good! some minor changes requested. Haven't run a resulting script on Baskerville though and will leave that to you to check.

I also think some of the tests are breaking right now so worth rebasing to develop so you pick up latest fixes

File: src/modsim2/model/wandb_logger.py (hunk 2 above)
Comment (philswatton) [reply src/modsim2/model/wandb_logger.py:3]: Done

File: configs/trainer.yaml (hunk 3 above)
Comment (philswatton) [reply configs/trainer.yaml:2]: Yes - looking at the wandb plots val_loss seems to plateau after about 4-6 epochs at most. Could lower this now or leave for #26?

File: scripts/train_all.sh (hunk 4 above)
Comment (philswatton) [reply scripts/train_all.sh:1]: Done

File: scripts/templates/train-template.sh (hunk 5 above)
Comment (philswatton) [reply scripts/templates/train-template.sh:1]: done

File: scripts/templates/train-template.sh (hunk 7 above)
Comment (philswatton) [reply scripts/templates/train-template.sh:18]: no - this is a leftover from the boilerplate in the tutorial, I've removed it now

File: scripts/templates/train-template.sh (hunk 8 above)
Comment (philswatton) [reply scripts/templates/train-template.sh:2]: Done

File: src/modsim2/model/wandb_logger.py (hunk 9 above)
Comment (philswatton) [reply src/modsim2/model/wandb_logger.py:78]: This is a c/p from the pl source code. My understanding is `metadata` becomes `None` if not (l 79). Either way the metadata is passed as an argument to `wandb.Artifact()` where `None` is the default argument

File: scripts/templates/train-template.sh (hunk 10 above)
Comment (philswatton) [reply scripts/templates/train-template.sh:19]: done

File: scripts/templates/train-template.sh (hunk 8 above)
Comment (philswatton) [reply scripts/templates/train-template.sh:2]: done

File: scripts/train_models.py (hunk 1 above)
Comment (philswatton) [reply scripts/train_models.py:50]: As far as I can tell no - passing `none` to a datamodule for the `seed` argument will cause an error when `.setup()` is called, even after running `seed_everything`. Looking at the source code it seems the datamodule needs its own seen, and doesn't draw on a global seed at all.

File: scripts/README.md (hunk 6 above)
Comment (philswatton) [reply scripts/README.md:31]: Have added these bits to the README

File: scripts/train_models.py (hunk 1 above)
Comment (lannelin) [reply scripts/train_models.py:50]: ah ok, that makes sense. Good stuff. Were you able to verify that `seed_everything` gives desired consistency for training runs?

File: configs/trainer.yaml (hunk 3 above)
Comment (lannelin) [reply configs/trainer.yaml:2]: Cool. Happy to leave for 26

File: src/modsim2/model/wandb_logger.py (hunk 9 above)
Comment (lannelin) [reply src/modsim2/model/wandb_logger.py:78]: ok happy to leave as is

File: scripts/README.md [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # Scripts
```
Comment (lannelin) [inline scripts/README.md:1]: there's a small typo in this file on wandb as "wadnb", sorry not strictly from this PR

File: scripts/README.md [hunk 12]
```diff
@@ -28,16 +28,33 @@ Call:
...
+Alternatively, you can call each of the individual scripts individually, e.g.
 
 ```bash
-trainer_scripts/drop-only_0_0_trainer.sh
+sbatch trainer_scripts/drop-only_0_0_trainer.sh
 ```
 
 to train models for that particular DMPair. Note that logging is performed with wadnb and so you will need this set up on your machine in order to proceed.
 
+### Requirements
```
Comment (lannelin) [inline scripts/README.md:47]: either put this at the start of the section or link to it at the start (e.g. "for prerequisite steps see [requirements](#dummylink))

**Review by lannelin — APPROVED** [review]
looks good! Approved with some very minor changes requested.
Could you also verify a training run (and consistency given by seeding) before merging?

File: scripts/README.md (hunk 12 above)
Comment (philswatton) [reply scripts/README.md:49]: Done

File: scripts/README.md (hunk 11 above)
Comment (philswatton) [reply scripts/README.md:1]: fixed!

**Comment by philswatton** [comment]
I added an extra arg to `Trainer` to make training deterministic given the seed - all the rest is done too!
